sdks/go/pkg/beam/util/diagnostics/diagnostics.go

```diff
@@ -0,0 +1,68 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package diagnostics
+
+import (
+	"bufio"
+	"context"
+	"os"
+
+	"runtime/debug"
+
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/io/filesystem"
+	_ "github.com/apache/beam/sdks/v2/go/pkg/beam/io/filesystem/gcs"
+)
+
+func UploadHeapDump(ctx context.Context, dest string, additionalDataToWrite string) error {
+	heapDumpLoc := "heapDump"
+
+	f, err := os.Create(heapDumpLoc)
+	if err != nil {
+		return err
+	}
+	debug.WriteHeapDump(f.Fd())
```

Comment:
I'm not sure if this is getting what we want now that I'm looking into this.

1. It's not clear to me that we're actually getting the heap from the spawned process. So what are we actually getting? The example heap dump is only ~8MB, which feels closer to what the launcher process is probably using, rather than the worker on death...
2. AFAICT we might not actually want a debug.WriteHeapDump dump generally anyway. It's the whole heap. Every Byte. Attempting to actually read the linked example dump has been... difficult. https://pkg.go.dev/runtime/pprof#WriteHeapProfile is likely easier for most folks to deal with and will indicate where the leaks are coming from if any.

Reply:
I think you're right. I got my wires crossed as I moved this code around and thought it wasn't showing up because we never successfully allocate the array that causes the OOM. On closer inspection, even if we successfully allocate a large array in advance, it still won't show up correctly. 

Just about the only path forward at this point would seem to be monitoring memory usage from a separate process and uploading a heap profile when we get close to full consumption. That's probably expensive though, and its not 100% clear to me we want to do that at the cost of performance. We at least will want to gate it behind an experiment.

I'll think a little bit harder about this one

Reply:
I changed this to periodically collect a profile from the worker process. When the process exits with an error, boot.go checks for the existence of a profile and uploads it if it exists.